src/js/jsx/tools/GuidesOverlay.jsx

```diff
@@ -0,0 +1,251 @@
+
+        /**
+         * Keeps track of current mouse position so we can rerender the overlaid layers correctly
+         * @type {number}
+         */
+        _currentMouseX: null,
+
+        /**
+         * Keeps track of current mouse position so we can rerender the overlaid layers correctly
+         * @type {number}
+         */
+        _currentMouseY: null,
+
+        /**
+         * Owner group for all the overlay svg elements
+         *
+         * @type {SVGElement}
+         */
+        _scrimGroup: null,
+
+        /**
+         * Debounced draw function, for performance
+         * 
+         * @type {function}
+         */
+        _drawDebounced: null,
+
+        getStateFromFlux: function () {
+            var flux = this.getFlux(),
+                applicationStore = flux.store("application"),
+                toolStore = flux.store("tool"),
+                modalState = toolStore.getModalToolState(),
+                currentTool = toolStore.getCurrentTool(),
+                currentDocument = applicationStore.getCurrentDocument();
+
+            return {
+                document: currentDocument,
+                tool: currentTool,
+                modalState: modalState
+            };
+        },
+
+        componentWillMount: function () {
+            this._drawDebounced = _.debounce(this.drawOverlay, DEBOUNCE_DELAY);
+        },
+
+        componentWillUnmount: function () {
+            OS.removeListener("externalMouseMove", this.mouseMoveHandler);
+        },
+
+        componentDidMount: function () {
+            this._currentMouseX = null;
+            this._currentMouseY = null;
+            
+            this._drawDebounced();
+            
+            // Marquee mouse handlers
+            OS.addListener("externalMouseMove", this.mouseMoveHandler);
+        },
+
+        componentDidUpdate: function () {
+            // Redraw immediately when we're in a modal state
+            if (this.state.modalState) {
+                this.drawOverlay();
+            } else {
+                this._drawDebounced();
+            }
+        },
+
+        /**
+         * Attaches to mouse move events coming from Photoshop
+         * so we can set highlights manually. We have to resort to this
+         * because external mouse move events do not cause :hover states
+         * in DOM elements
+         *
+         * @private
+         * @param {CustomEvent} event EXTERNAL_MOUSE_MOVE event coming from _spaces.OS
+         */
+        mouseMoveHandler: function (event) {
+            if (this.isMounted()) {
+                this._currentMouseX = event.location[0];
+                this._currentMouseY = event.location[1];
+                
+                this.updateMouseOverHighlights();
+            }
+        },
+
+        /**
+         * Calls all helper functions to draw sampler overlay
+         * Cleans it first
+         * @private
+         */
+        drawOverlay: function () {
```

Comment:
Fair.  This might suggest that it would breathe a little better if it were factored out into a separate module.  food for thought.  not for now.